substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -995,15 +995,21 @@ public int modifiers() {
     }
 
     private static boolean filterLineInfoSourceMapping(SourceMapping sourceMapping) {
-        if (!SubstrateOptions.OmitInlinedMethodDebugLineInfo.getValue()) {
-            return true;
+        NodeSourcePosition sourcePosition = sourceMapping.getSourcePosition();
+        /* Don't report line info for zero length ranges. */
+        if (sourceMapping.getStartOffset() == sourceMapping.getEndOffset()) {
+            return false;
```

Comment:
I am not at all sure why we see ranges with zero length in the source mappings. They appear always to be followed by a non-zero length range with the same start address so gdb certainly ignores any line info they present in favour of that presented by the successor range. In consequence it seems to make sense just to drop them as far as line and inline info is concerned. However, it may be that this is not correct when it comes to identifying local var info.

@olpaw perhaps you could provide an explanation of why these zero length ranges are recorded and whether droppping them is the correct approach.

Reply:
I debugged this a bit and i can see that 95% of those SourceMappings were created from `org.graalvm.compiler.lir.StandardOp.LabelOp` `LIRInstruction`s. For those, imo, it makes sense that they have zero length range. After all it's just Labels generated by the `LIRGenerator`. They are synthetic and do not originate from the source in a meaningful way. The other empty SourceMappings I see are caused by 

- `com.oracle.svm.core.graal.amd64.SubstrateAMD64Backend.DeadEndOp` nodes
- org.graalvm.compiler.lir.amd64.AMD64Move$MoveToRegOp nodes which seem to be NOPs (e.g. `rax|QWORD[.] = MOVE rax`.
- `org.graalvm.compiler.lir.StandardOp$JumpOp` Nodes (mostly from `B_n` to `B_n+1` but not always)

`DeadEndOp` nodes are similar to `LabelOp` (generated by the backend not relatable to source code).

All in all I consider it safe to ignore all those SourceMappings since they do not provide any information that we are interested in during debugging. @christianwimmer @peter-hofer

Reply:
Below the patch I used to gather the above information
```patch
diff --git a/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/CompilationResult.java b/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/CompilationResult.java
index 81dcedc609..6c861adef7 100644
--- a/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/CompilationResult.java
+++ b/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/CompilationResult.java
@@ -41,6 +41,7 @@ import org.graalvm.collections.EconomicSet;
 import org.graalvm.collections.Equivalence;
 import org.graalvm.compiler.core.common.CompilationIdentifier;
 import org.graalvm.compiler.graph.NodeSourcePosition;
+import org.graalvm.compiler.lir.LIRInstruction;
 import org.graalvm.compiler.options.OptionValues;
 import org.graalvm.compiler.serviceprovider.GraalServices;
 
@@ -611,7 +612,7 @@ public class CompilationResult {
         infopoints.add(infopoint);
     }
 
-    public void recordSourceMapping(int startOffset, int endOffset, NodeSourcePosition sourcePosition) {
+    public void recordSourceMapping(int startOffset, int endOffset, LIRInstruction sourcePosition) {
         checkOpen();
         sourceMapping.add(new SourceMapping(startOffset, endOffset, sourcePosition));
     }
diff --git a/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/SourceMapping.java b/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/SourceMapping.java
index d81e0b2fe0..a3f775c621 100644
--- a/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/SourceMapping.java
+++ b/compiler/src/org.graalvm.compiler.code/src/org/graalvm/compiler/code/SourceMapping.java
@@ -27,6 +27,7 @@ package org.graalvm.compiler.code;
 import static jdk.vm.ci.meta.MetaUtil.identityHashCodeString;
 
 import org.graalvm.compiler.graph.NodeSourcePosition;
+import org.graalvm.compiler.lir.LIRInstruction;
 
 /**
  * This provides a mapping between a half-open range of PCs in the generated code and a
@@ -40,11 +41,13 @@ public final class SourceMapping {
     private final int endOffset;
 
     private final NodeSourcePosition sourcePosition;
+    public final LIRInstruction op;
 
-    public SourceMapping(int startOffset, int endOffset, NodeSourcePosition sourcePosition) {
+    public SourceMapping(int startOffset, int endOffset, LIRInstruction op) {
         this.startOffset = startOffset;
         this.endOffset = endOffset;
-        this.sourcePosition = sourcePosition;
+        this.sourcePosition = op.getPosition();
+        this.op = op;
     }
 
     public int getStartOffset() {
diff --git a/compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/asm/CompilationResultBuilder.java b/compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/asm/CompilationResultBuilder.java
index ac4e5c0177..9604632e76 100644
--- a/compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/asm/CompilationResultBuilder.java
+++ b/compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/asm/CompilationResultBuilder.java
@@ -345,7 +345,7 @@ public class CompilationResultBuilder {
         compilationResult.recordInfopoint(pos, debugInfo, reason);
     }
 
-    public void recordSourceMapping(int pcOffset, int endPcOffset, NodeSourcePosition sourcePosition) {
+    public void recordSourceMapping(int pcOffset, int endPcOffset, LIRInstruction sourcePosition) {
         compilationResult.recordSourceMapping(pcOffset, endPcOffset, sourcePosition);
     }
 
@@ -592,7 +592,10 @@ public class CompilationResultBuilder {
             int start = asm.position();
             op.emitCode(this);
             if (op.getPosition() != null) {
-                recordSourceMapping(start, asm.position(), op.getPosition());
+                if (start == asm.position()) {
+                    System.out.println("Created empty SourceMapping at " + start + " caused by node " + op.getClass() + " " + op);
+                }
+                recordSourceMapping(start, asm.position(), op);
             }
             if (LIR_INSTRUCTION_VERIFIERS.size() > 0 && start < asm.position()) {
                 int end = asm.position();
diff --git a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java
index 5e6ed24b63..ae46595497 100644
--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java
@@ -998,6 +998,7 @@ class NativeImageDebugInfoProvider implements DebugInfoProvider {
         NodeSourcePosition sourcePosition = sourceMapping.getSourcePosition();
         /* Don't report line info for zero length ranges. */
         if (sourceMapping.getStartOffset() == sourceMapping.getEndOffset()) {
+            System.out.printf("Zero-length LIRInstruction %s: %s, SourceMapping %s%n", sourceMapping.op.getClass(), sourceMapping.op, sourcePosition);
             return false;
         }
         /* Don't report inline line info unless the user has configured it. */
```

Reply:
Ok, so it seems this is doig the right thing.